### PR TITLE
Add support for longer Datums

### PR DESCRIPTION
**Pull request by rustamlabs**

### CODE REVIEW

File: src/types/internal.ts [hunk 1]
```diff
@@ -365,23 +364,20 @@ export type OutputDestination = {
...
-export type ParsedOutputBabbageMap = {
-    type: TxOutputType.MAP_BABBAGE;
+export type ParsedOutput = {
+    type: TxOutputType;
     amount: Uint64_str;
     tokenBundle: ParsedAssetGroup<Uint64_str>[];
     destination: OutputDestination;
-    datum?: Datum;
+    datum: ParsedDatum | null;
     scriptHex?: string | null;
```
Comment (janmazak) [inline src/types/internal.ts:381]:
```suggestion
    scriptHex?: HexString | null;
```

**Review by janmazak — COMMENTED** [review]
Some small things to fix quickly.

File: src/interactions/signTx.ts [hunk 2]
```diff
@@ -142,48 +144,63 @@ function* signTx_addOutput(
...
+    if (output.datum) {
+        yield send({
+            p1: P1.STAGE_OUTPUTS,
+            p2: P2.DATUM_OPTION,
+            data: serializeTxOutputDatum(output, version),
+            expectedResponseLength: 0,
+        })
+        // Datum Chunks
+        if (output.datum.type === DatumType.INLINE
+            && output.datum.datumHex.length / 2 > MAX_DATUM_CHUNK_SIZE) {
```
[the remark this answers is not in the document]
Comment (rustamlabs) [reply src/interactions/signTx.ts:174]: Well, that does not compile:
` Property 'datumHex' does not exist on type 'ParsedDatum'`

[the remark this answers is not in the document]
Comment (rustamlabs) [reply src/interactions/signTx.ts:174]: However I can add a comment describing it

File: src/parsing/transaction.ts [hunk 3]
```diff
@@ -127,6 +129,39 @@ function parseTokenBundle<T>(tokenBundle: AssetGroup[], emptyTokenBundleAllowed:
...
+            datum = null
+            break
+        }
+
+    } else { // Alonzo
+        datum = output.datumHashHex == null
+            ? null
+            : {
+                type: DatumType.HASH,
+                datumHashHex: parseHexStringOfLength(output.datumHashHex, SCRIPT_DATA_HASH_LENGTH, InvalidDataReason.SCRIPT_DATA_HASH_WRONG_LENGTH),
```
[the remark this answers is not in the document]
Comment (rustamlabs) [reply src/parsing/transaction.ts:159]: We need to know the output.type as well, if not enclosed withe type check, `output.datum?.type` throws `TS2339: Property 'datum' does not exist on type 'TxOutput'.   Property 'datum' does not exist on type  'TxOutputAlonzo'`
But I can create seperate method to parse datum hash

File: src/types/internal.ts (hunk 1 above)
Comment (rustamlabs) [reply src/types/internal.ts:381]: that has been done on #171

File: src/interactions/signTx.ts [hunk 4]
```diff
@@ -142,48 +144,63 @@ function* signTx_addOutput(
...
+    }
 
-  // Basic data
-  yield send({
-      p1: P1.STAGE_OUTPUTS,
-      p2: P2.BASIC_DATA,
-      data: serializeTxOutputBasicParams(output, version),
-      expectedResponseLength: 0,
-  })
+    // Basic data
```
[the remark this answers is not in the document]
Comment (rustamlabs) [reply src/interactions/signTx.ts:154]: I  have seen that on other parts of code, but now I noticed mostly there is not. (signTx_addTokenBundle) I will take a look at that. Tab sizes and indentations are not consistent either, it is different on every place.

**Pull request merged**